### PR TITLE
[#110] Include FA2 recent changes to update_operators parameter.

### LINKED ISSUE #110: Incorporate "FA2 operators per token type" changes
# Clarification and motivation

FA2 was updated recently. See https://gitlab.com/tzip/tzip/-/merge_requests/85.

# Acceptance criteria

Stablecoin contract complies with FA2 version that took place after https://gitlab.com/tzip/tzip/-/merge_requests/85 was merged. Specifically, `update_operators` is updated.

### PR DESCRIPTION
**Pull request by sras**
## Description

Include change in https://gitlab.com/tzip/tzip/-/merge_requests/85 w.r.t `update_operators`.

## Related issue(s)

Resolves #110 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [x] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: haskell/test/SMT.hs [hunk 1]
```diff
@@ -406,8 +409,8 @@ generateTokenOwnerAction idx = do
       operator <- getRandomOperator
       owner <- getRandomOwner
       updateOperation <- lift $ elements
-        [ FA2.Add_operator (#owner .! owner, #operator .! operator)
-        , FA2.Remove_operator (#owner .! owner, #operator .! operator)
+        [ FA2.Add_operator (#owner .! owner, (#operator .! operator, #token_id .! 0))
+        , FA2.Remove_operator (#owner .! owner, (#operator .! operator, #token_id .! 0))
```
Comment (gromakovsky) [inline haskell/test/SMT.hs:413]: Why not generate an action with invalid token_id and check that it's rejected?

Comment (sras) [reply haskell/test/SMT.hs:414]: Included.

File: haskell/test-common/Lorentz/Contracts/Test/Common.hs [hunk 2]
```diff
@@ -224,3 +225,6 @@ mkInitialStorage OriginationParams{..} = let
       -> Map (Address, (Address, Natural)) ()
       -> Map (Address, (Address, Natural)) ()
     foldFn ow ops m = foldr (\a b -> Map.insert (ow, (a, 0)) () b) m ops
+
+mgmContractPaused :: ExecutorError -> IntegrationalScenario
+mgmContractPaused = lExpectFailWith (== [mt|CONTRACT_PAUSED|])
```
Comment (gromakovsky) [inline haskell/test-common/Lorentz/Contracts/Test/Common.hs:230]: Why did you move it btw?

File: haskell/test/Lorentz/Contracts/Test/FA2.hs [hunk 3]
```diff
@@ -579,6 +579,34 @@ fa2Spec fa2Originate = do
...
+
+          withSender wallet1 $ do
+            let operatorParam =
+                  (#owner .! wallet1, (#operator .! commonOperator, #token_id .! 0))
+
+            err <- expectError $  lCallEP fa2contract (Call @"Update_operators")
+                    [Add_operator operatorParam]
+            mgmContractPaused err
+
+    it "remove operator call validates token id" $
```
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/FA2.hs:596]: Is this description correct?

File: docs/specification.md [hunk 4]
```diff
@@ -360,11 +360,15 @@ Parameter (in Michelson)
   (or
     (pair %add_operator
       (address %owner)
-      (address %operator)
+      (pair
+        (address %operator)
+        (nat %token_id))
```
Comment (gromakovsky) [inline docs/specification.md:365]: Apparently you forgot to update pseudocode definitions above.
```
operator_param =
  ( address         :owner
  , address         :operator
  )
```

File: ligo/stablecoin/spec.ligo [hunk 5]
```diff
@@ -288,7 +289,7 @@ type owner is address
 type operator is address
 
 type operators is
-  big_map ((owner * operator), unit)
+  big_map ((owner * (operator * token_id)), unit)
```
Comment (gromakovsky) [inline ligo/stablecoin/spec.ligo:292]: What's the point to store `token_id` here? Isn't it always 0?

Comment (gromakovsky) [reply ligo/stablecoin/spec.ligo:292]: This construct makes sense for multiasset contracts, but for singleasset (like stablecoin) it looks redundant.

File: haskell/test-common/Lorentz/Contracts/Test/Common.hs (hunk 2 above)
Comment (sras) [reply haskell/test-common/Lorentz/Contracts/Test/Common.hs:230]: @gromakovsky To use in the `Lorentz.Contracts.Test.FA2` module to test that `update_operator` is pausable.

Comment (gromakovsky) [reply haskell/test-common/Lorentz/Contracts/Test/Common.hs:230]: Ah, I thought it previously was in the library. Ok then.

**Comment by gromakovsky** [comment]
Looks good now, but there is a ton of conflicts.

**Comment by gromakovsky** [comment]
> I updated changelog unless I am sure my changes are
not essential.

Reminder.

**Comment by sras** [comment]
@gromakovsky I have 

1. rebased, 
2. removed is_operator entrypoint and updated tests
3. updated spec and changelog.

File: haskell/test/Lorentz/Contracts/Test/FA2.hs [hunk 6]
```diff
@@ -471,78 +471,69 @@ fa2Spec fa2Originate = do
 
   -- These tests require permission descriptor to be configured that allows for operator transfer.
   -- We have such a configuration set by default in defaultOriginationParams.
+  let
+    checkForOperator stablecoinContract owner operator expectation =
+      lExpectStorage stablecoinContract $ \storage -> do
+        let found = isJust $ M.lookup (owner, operator) (unBigMap $ sOperators storage)
+        if found /= expectation
+            then Left $ CustomTestError $ "Unexpected operator status. Expected: " <> (show expectation) <> " Found: " <> (show found)
```
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/FA2.hs:479]:
```suggestion
            then Left $ CustomTestError $ "Unexpected operator status. Expected: " <> show expectation <> " Found: " <> show found
```

Comment (gromakovsky) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs:479]: Also please split it into two lines, this one looks too long.

**Review by gromakovsky — APPROVED** [review]
LGTM except for one minor thing. Feel free to merge after cleaning up commits.

File: haskell/test/Lorentz/Contracts/Test/FA2.hs (hunk 6 above)
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs]: > Also please split it into two lines, this one looks too long.

Have changed.
